src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+            CancellationToken cancellationToken)
+        {
+            var actionsBuilder = new ArrayBuilder<CodeAction>();
+            var (isParameterized, hasOptionalParameter) = await ShouldExpressionDisplayCodeActionAsync(document, expression, methodSymbol, cancellationToken).ConfigureAwait(false);
+            if (isParameterized)
+            {
+                actionsBuilder.Add(CreateNewCodeAction(FeaturesResources.Introduce_parameter_for_0,
+                    document, expression, methodSymbol, containingMethod, allOccurrences: false, trampoline: false, overload: false, cancellationToken));
+                actionsBuilder.Add(CreateNewCodeAction(FeaturesResources.Introduce_parameter_for_all_occurrences_of_0,
+                    document, expression, methodSymbol, containingMethod, allOccurrences: true, trampoline: false, overload: false, cancellationToken));
+
+                if (!hasOptionalParameter && methodSymbol.MethodKind is not MethodKind.Constructor)
+                {
+                    actionsBuilder.Add(CreateNewCodeAction(
+                        FeaturesResources.Extract_method_to_invoke_at_all_callsites_for_0, document, expression,
+                        methodSymbol, containingMethod, allOccurrences: false, trampoline: true, overload: false, cancellationToken));
+                    actionsBuilder.Add(CreateNewCodeAction(
+                        FeaturesResources.Extract_method_to_invoke_at_all_callsites_for_all_occurrences_of_0, document,
+                        expression, methodSymbol, containingMethod, allOccurrences: true, trampoline: true, overload: false, cancellationToken));
+
+                    actionsBuilder.Add(CreateNewCodeAction(
+                        FeaturesResources.Introduce_overload_with_new_parameter_for_0, document,
+                        expression, methodSymbol, containingMethod, allOccurrences: false, trampoline: false, overload: true, cancellationToken));
+                    actionsBuilder.Add(CreateNewCodeAction(
+                         FeaturesResources.Introduce_overload_with_new_parameter_for_all_occurrences_of_0, document,
+                         expression, methodSymbol, containingMethod, allOccurrences: true, trampoline: false, overload: true, cancellationToken));
+                }
+            }
+
+            return actionsBuilder.ToImmutable();
+
+            // Local function to create a code action with more ease
+            MyCodeAction CreateNewCodeAction(string actionName, Document document,
+                TExpressionSyntax expression, IMethodSymbol methodSymbol, SyntaxNode containingMethod,
+                bool allOccurrences, bool trampoline,
+                bool overload, CancellationToken cancellationToken)
+            {
+                return new MyCodeAction(actionName, c => IntroduceParameterAsync(
+                    document, expression, methodSymbol, containingMethod, allOccurrences, trampoline, overload, cancellationToken));
```

Comment:
this will crash badly in practice.  you are capturing a cancellation token across a lambda invoked at an entirely different time.  you need to pass in `c` here instead.

Reply:
fixed